contrib/make_training_pairs.py

```diff
@@ -1,93 +1,154 @@
+            self.connected.cache_clear()
+            yield edge
+
+
+def lazy_resolve(view: View, resolver: Resolver, id: Identifier) -> Optional[Entity]:
+    """
+    Get an entity merging all connected entities from the view.
+    """
+    cluster: Optional[Entity] = None
+
+    for ident in resolver.connected(id):
+        entity = view.get_entity(ident.id)
+        if entity is None:
             continue
-        for ds in datasets.get(ref, []):
-            yield ref, ds
-
-
-def get_partial(
-    resolver: Resolver, store: Store, spec: Tuple[str, Dataset]
-) -> Optional[Entity]:
-    id, ds = spec
-    loader = store.view(ds, external=True)
-    canonical = resolver.get_canonical(id)
-    entity = loader.get_entity(canonical)
-    if entity is None:
-        return None
-    entity.id = id
-    return entity
-
-
-def export_training_pairs(scope: Dataset):
-    resolver = get_resolver()
-    catalog = get_catalog()
-    datasets: Dict[str, Set[Dataset]] = defaultdict(set)
-    with engine_read() as conn:
-        for entity_id, ds in entities_datasets(conn, scope):
-            if ds not in scope.leaf_names:
-                continue
-            dsa = catalog.get(ds)
-            if dsa is not None:
-                datasets[entity_id].add(dsa)
-
-    log.info("Loaded %d entity ID mappings..." % len(datasets))
-    pairs: Dict[Tuple[Tuple[str, Dataset], Tuple[str, Dataset]], Judgement] = {}
-    judgements: Dict[Judgement, int] = defaultdict(int)
-    for canonical_id in resolver.canonicals():
-        parts = list(get_parts(resolver, datasets, canonical_id))
-        for left, right in combinations(parts, 2):
-            left, right = max(left, right), min(left, right)
-            pairs[(left, right)] = Judgement.POSITIVE
-            judgements[Judgement.POSITIVE] += 1
-        for edge in resolver.nodes[canonical_id]:
-            if edge.judgement in (Judgement.NEGATIVE, Judgement.UNSURE):
-                source_canonical = resolver.get_canonical(edge.source)
-                other = edge.target if source_canonical == canonical_id else edge.source
-                for other_part in get_parts(resolver, datasets, other):
-                    for part in parts:
-                        part, other_part = max(part, other_part), min(part, other_part)
-                        # pairs[(part, other_part)] = edge.judgement
-                        # Export unsure as negative:
-                        pairs[(part, other_part)] = edge.judgement
-                        judgements[edge.judgement] += 1
+        if cluster is None:
+            cluster = entity
+        else:
+            cluster.merge(entity)
+    return cluster
+
+
+def generate_pairs(
+    resolver_path: Path,
+    dataset: Dataset,
+) -> Generator[Tuple[Entity, Entity, Judgement], None, None]:
+    chrono_resolver = ChronoResolver(resolver_path)
+    store = get_store(dataset, Linker({}))
+    store.sync()
+    view = store.view(dataset, external=False)
```

Comment:
Oh this is kind of interesting: a good number of the negative/unsure judgements should be `external`-based, and then the external fragment isn't re-stored when the enricher runs again. So we lose these.